src/test/com/fulcrologic/fulcro/algorithms/normalized_state_helpers_spec.cljc

```diff
@@ -0,0 +1,329 @@
+                                   :person/children [[:person/id 3]
+                                                     [:person/id 4]
+                                                     [:person/id 5]]}
+                                3 {:person/name "person-3"}
+                                4 {:person/name     "person-4"
+                                   :person/spouse   [:person/id 6]
+                                   :person/children [:person/id 7]}
+                                5 {:person/name "person-5"}
+                                6 {:person/id       6
+                                   :person/name     "person-6"
+                                   :person/spouse   [:person/id 4]
+                                   :person/children [:person/id 7]}
+                                7 {:person/name "person-7"}}
+                 :car/id       {1 {:car/model  "model-1"
+                                   :car/engine [:engine/id 1]}
+                                2 {:car/model  "model-2"
+                                   :car/engine [:engine/id 2]}}
+                 :engine/id    {1 {:engine/name "engine-1"}
+                                2 {:engine/name "engine-2"}}
+                 :email/id     {1 {:email/provider "Google"}
+                                2 {:email/provider "Microsoft"}}}]
+      (assertions
+
+        "Removes a single, to-one and non-recursive cascased entity"
+        (-> (nsh/remove-entity* state [:person/id 1] #{:person/email})
+            (get [:email/id 1])
+            nil?) => true
+
+        "Removes a single, to-many and non-recursive cascased entity"
+        (-> (nsh/remove-entity* state [:person/id 1] #{:person/cars})
+            (affects...
+              (get [:car/id 1])
+              (get [:car/id 2]))
+            nil?) => true
+
+        "Removes multiple, to-one and non-recursive cascased entity"
+        (-> (nsh/remove-entity* state [:person/id 1] #{:person/email :person/cars})
+            (affects...
+              (get [:email/id 1])
+              (get [:car/id 1])
+              (get [:car/id 2]))
+            nil?) => true
+
+        "Removes a single, to-many and recursive cascased entities"
+        (-> (nsh/remove-entity* state [:person/id 1] #{:person/children})
+            (affects...
+              (get [:person/id 3])
+              (get [:person/id 4])
+              (get [:person/id 5])
+              (get [:person/id 7]))
+            nil?) => true
+
+        "Removes multiple, to-many and recursive cascased entities"
+        (-> (nsh/remove-entity* state [:person/id 1] #{:person/children :person/spouse})
+            (affects...
+              (get [:person/id 2])
+              (get [:person/id 3])
+              (get [:person/id 4])
+              (get [:person/id 5])
+              (get [:person/id 6])
+              (get [:person/id 7]))
+            nil?) => true
+
+        ;; TODO add cases for person -> car -> engine
```

Comment:
@awkay, so these are the test cases which I could think of, given the discussions we've had - please do give these a look.

Reply:
I've shrunk the tests cases and separated the behaviors for focused test data